catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java

```diff
@@ -126,104 +216,196 @@ public boolean onOptionsItemSelected(MenuItem item) {
+			nameInputLayout.getEditText().removeTextChangedListener(nameInputTextWatcher);
+			nameInputLayout.setEnabled(false);
+			descriptionInputLayout.setEnabled(false);
+			return;
+		}
 
-		nameInputLayout.getEditText().setText(currentProject.getName());
-		descriptionInputLayout.getEditText().setText(currentProject.getDescription());
+		setShowProgressBar(true);
+		showSelectTagsDialog();
+	}
 
-		nameInputLayout.getEditText().addTextChangedListener(textChangedListener);
-		invalidateOptionsMenu();
+	private void showSelectTagsDialog() {
+		final List<String> checkedTags = new ArrayList<>();
+		final String[] availableTags = tags.toArray(new String[tags.size()]);
+
+		new AlertDialog.Builder(this)
+				.setTitle(R.string.upload_tag_dialog_title)
+				.setMultiChoiceItems(availableTags, null, new DialogInterface.OnMultiChoiceClickListener() {
+					@Override
+					public void onClick(DialogInterface dialog, int indexSelected, boolean isChecked) {
+						if (isChecked) {
+							if (checkedTags.size() >= MAX_NUMBER_OF_TAGS_CHECKED) {
+								((AlertDialog) dialog).getListView().setItemChecked(indexSelected, false);
+							} else {
+								checkedTags.add(availableTags[indexSelected]);
+							}
+						} else {
+							checkedTags.remove(availableTags[indexSelected]);
+						}
+					}
+				})
+				.setPositiveButton(getText(R.string.next), new DialogInterface.OnClickListener() {
+					@Override
+					public void onClick(DialogInterface dialog, int id) {
+						project.setTags(checkedTags);
+						showProgressDialogAndUploadProject();
+					}
+				})
+				.setNegativeButton(getText(R.string.cancel), new DialogInterface.OnClickListener() {
+					@Override
+					public void onClick(DialogInterface dialog, int id) {
+						Utils.invalidateLoginTokenIfUserRestricted(ProjectUploadActivity.this);
+						setShowProgressBar(false);
+						setNextButtonEnabled(true);
+					}
+				})
+				.setCancelable(false)
+				.show();
 	}
 
-	private void onNextButtonClick() {
-		String name = nameInputLayout.getEditText().getText().toString().trim();
+	private void showProgressDialogAndUploadProject() {
+		final String name = nameInputLayout.getEditText().getText().toString().trim();
 		String description = descriptionInputLayout.getEditText().getText().toString().trim();
 
-		String error = textChangedListener.validateInput(name);
-
-		if (error != null) {
-			nameInputLayout.setError(error);
-			invalidateOptionsMenu();
-			return;
+		if (!project.getName().equals(name)) {
+			ProjectRenameTask
+					.task(project.getName(), name);
+			ProjectLoadTask
+					.task(name, this);
+			project = ProjectManager.getInstance().getCurrentProject();
 		}
 
-		Project project = ProjectManager.getInstance().getCurrentProject();
+		project.setDescription(description);
+		project.setDeviceData(this);
+
+		File projectDir = new File(DEFAULT_ROOT_DIRECTORY, project.getName());
 
-		if (Utils.isDefaultProject(project, this)) {
```

Comment:
are we losing this functionality at this point? is there a check somewhere else if it's the default project?

Reply:
yes! you're absolutely right! I missed that one. thanks :wink: